#234 support for references for arrays

243826 opened the pull request:
Fixed Issue #234 by resolving the unresolved references.

243826 commented:
Pardon my whitespace changes. It seems that the file was badly formatted and my editor reformatted it even before i started making changes. I kept the changes since it looked like the changes were for the betterment...

but open to suggestions.

stevehu commented:
@243826 Thanks for the contribution. No worry about the whitespaces. As we are not using array references, I am wondering if you could add one or more unit tests to demo how to use it and these test can greatly help us to understand your code.

jiachen1120 commented on light-rest-4j/src/main/java/com/networknt/codegen/rest/OpenApiGenerator.java:
```diff
@@ -273,48 +286,123 @@ public void generate(String targetPath, Object model, Any config) throws IOExcep
+                            };
+                        } else {
+                            r = () -> {
+                                final int referencesCount = references.size();
```
If this line is only for testing, can it be commented out?

243826 replied:
It's not for testing... looks like while refactoring I missed out some code. I'll either add that code or remove this line.

243826 replied:
This one is addressed.

jiachen1120 commented on light-rest-4j/src/main/java/com/networknt/codegen/rest/OpenApiGenerator.java:
```diff
@@ -273,48 +286,123 @@ public void generate(String targetPath, Object model, Any config) throws IOExcep
+                            throw new IllegalStateException("Properties empty for " + classVarName + "!");
+                        }
+
+                        references.put(classVarName, props.get(0).get("type"));
```
If I understand correctly, if only store type information in the refenreces, would this ignore the enum case?

For example, if the specification contains following schema
```
error:
      type: object
      properties:
        code:
          $ref: '#/components/schemas/code' 
code:
      type: string
      enum: [test, test1]
```
The code would be generated as 
```
private String code
```
However, it should be 
```
public enum CodeEnum {...}
```

243826 replied:
Thanks for the feedback... Part of the reason I missed that is I am not that familiar with the code. But now with your concrete example, I know what I need to review. Will fix it.

243826 replied:
This one is addressed too.

243826 commented:
> @243826 Thanks for the contribution. No worry about the whitespaces. As we are not using array references, I am wondering if you could add one or more unit tests to demo how to use it and these test can greatly help us to understand your code.

yes, I think I should... I was just being lazy:

1. Since the openapi.json already has array of NotificationPreference, it's a positive test.
2. I did not spend enough time to understand how your tests validate that the generated code is sane. Do you have pointers to understand it quickly?

243826 commented:
> Hi @243826 : my current client has array references in a number of OpenAPI specs.
> 
> Could you please add a test which employs a (simple) OpenAPI spec showcasing arrays?
> 
> Thank you

Is it possible for you to share a stripped down version of the openapi3 specification? It will save me valuable time to think of an example which most likely will be madeup and meaningless.

stevehu commented:
@243826 We don't have end-to-end verification for the codegen at the moment in light-codegen; however, we have a lot of examples in light-example-4j and specification/config in [model-config](https://github.com/networknt/model-config) repo. Each time we do a build, we are testing some of the examples end-to-end in [light-bot](https://github.com/networknt/light-config-test/blob/master/light-bot/develop-build/build-test-all/config/develop-build.yml#L143). We can easily expand that to generate some example applications from specifications and start the server to test against the example response. I will open an issue in light-bot.

243826 commented:
I have added a unit test which validates the affected code generation. I also addressed the enum side of the story. Please review and let me know.

Also just to share with you guys - I am re-architecting a large code base and had laid out principles as to what the new architecture should support. I was mainly looking for a fast microservices framework  when I came across light-4j and was delighted to see that many of my principles match with how you guys laid out on your site. Keep up the good work. Let me know if you would like me to review new ideas. Since I am trying to use your framework, I am sure to come by to review the progress of this product.

stevehu commented:
@243826 Welcome to the team. I have sent you an invite to add to the contributor list. Please feel free to join any discussions and review any issues. Once the team members know which repo or area you are interested in, you will be added to the reviewer list by them :)

243826 commented:
@ddobrin I have addressed your change request as well. Could you validate it and/or approve the pull request so that it can be merged? Thanks.

243826 commented:
@stevehu Could you please look into merging this ASAP?

243826 commented:
what's the significance of develop branch vs master branch?

stevehu commented:
We release from the develop branch and then merge the develop to master. So that master branch is in sync with our latest release.

243826 commented:
[image: MailTag]does it make sense for me to then develop against the
develop branch instead of the master branch?

On Tue, Mar 5, 2019 at 3:09 PM Steve Hu <notifications@github.com> wrote:

> We release from the develop branch and then merge the develop to master.
> So that master branch is in sync with our latest release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/networknt/light-codegen/pull/238#issuecomment-469896569>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACk9w1tlLF5Qg-Fu_u4QwLy9nY4v9sz0ks5vTvkrgaJpZM4bVp3Y>
> .
>

stevehu commented:
Yes. All developers should work against the develop branch. Once the PR is submitted, we need to have two reviewers to approve the PR before merging into the develop branch for the next release. 

This is the contribution document. https://doc.networknt.com/contribute/development/
